sqlsorcery.py

```diff
@@ -0,0 +1,100 @@
+import pandas as pd
+import pandas_gbq
+from sqlalchemy import create_engine, Table, MetaData
+from sqlalchemy.engine.url import URL
+from sqlalchemy.exc import NoSuchTableError
+from sqlalchemy.schema import DropTable
+
+
+class Connection:
+    def __init__(
+        self,
+        dialect=None,
+        user=None,
+        pwd=None,
+        server=None,
+        port=None,
+        db=None,
+        schema=None,
+        creds=None,
+    ):
+        self.dialect = dialect
+        self.schema = schema
+        self.db = db
+        self.options = {
+            "username": user,
+            "password": pwd,
+            "host": server,
+            "port": port,
+            "database": db,
+        }
+        self.creds = creds
+        self.helpers = {}
+        self.engine = self._get_engine()
+
+    def _cstr(self):
+        return URL(self.dialect, **self.options)
+
+    def _bigquery_options(self):
+        if self.dialect == "bigquery":
+            self._get_creds()
+            self.helpers["credentials_path"] = self.creds
+
+    def _get_creds(self):
+        credentials = service_account.Credentials.from_service_account_file(self.creds)
+        pandas_gbq.context.credentials = credentials
+        pandas_gbq.context.project = self.db
+
+    def _get_engine(self):
+        self._bigquery_options()
+        return create_engine(self._cstr(), **self.helpers)
+
+    def query(self, sql_query, params=None):
+        if self.dialect == "bigquery":
+            return pandas_gbq.read_gbq(sql_query, progress_bar_type=None)
+        else:
+            return pd.read_sql_query(sql_query, con=self.engine, params=params)
+
+    def insert_into(self, table, df, if_exists="append", chunksize=None, dtype=None):
+        if self.dialect == "bigquery":
+            pandas_gbq.to_gbq(
+                dataframe=df,
+                destination_table=f"{self.schema}.{table}",
+                if_exists=if_exists,
+                chunksize=chunksize,
+                table_schema=dtype,
+            )
+        else:
+            df.to_sql(
+                name=table,
+                con=self.engine,
+                schema=self.schema,
+                if_exists=if_exists,
+                index=False,
+                chunksize=chunksize,
+                dtype=dtype,
+            )
+
+    def table(self, tablename):
+        return Table(
+            tablename,
+            MetaData(),
+            autoload=True,
+            autoload_with=self.engine,
+            schema=self.schema,
+        )
+
+    def drop(self, tablename):
+        try:
+            table = self.table(tablename)
+            self.engine.execute(DropTable(table))
+        except NoSuchTableError:
+            pass
+
+    def read_table(self, tablename):
+        if self.dialect == "bigquery":
+            return self.query(f"SELECT * FROM {self.schema}.{tablename}")
+        else:
+            return pd.read_sql_table(tablename, con=self.engine, schema=self.schema)
```

Comment:
I think this will error if the table doesn't exist. Either here or in the call-site we will need a try/except or a check that the table exists. If here, we probably want to at least return an indicator of whether it succeeded and log something. I don't have a good sense of best practice on whether we should just let the caller deal with errors or handle it here, but we should be consistent.

Reply:
@zkagin Not sure I understand the advantage of the try/except here. If this errors (and it will) if the table doesn't exist the job should throw an exception and exit. This is needed for pulling the course ids and if it fails then the endpoints that rely on that will also fail. Unless you are thinking it should just log the error but run all the endpoints that don't need course id? But that seems like it might hide the error in the mailer and you'd have to read the logs to see the failure.

Reply:
This is a good question perhaps of how we handle errors more broadly. I'm generally a proponent of graceful failure mostly because my experience is in consumer applications, but I can see how just throwing the error and crashing out makes it clear something is very wrong.

At the very least I understand the logic not to put the try/excepts in this library and to raise it to the call site. At the call site, you can then decide how to handle it, and there I'd support a try/except that logs the message, and then handles it appropriately (which in this case might be preventing any of the other course_id-dependent jobs from running). It'll make it more clear for anyone running it what happened and why, and potentially give them some methods for fixing it beyond parsing the raw error.